### PR TITLE
User defined averagines

### PR DESCRIPTION
**Pull request by poshul**
Added code to IsotopeDistribution to generate isotopic distributions with user specified elemental composition. Also added methods containing predefined compositions for DNA and RNA.

Added a parameter to FeatureFinderMultiplex to specify the type of averagine to use (from the presets in IsotopeDistribution) and optional parameters to MultiplexFiltering, MultiplexFilteringCentroided, and MultiplexFilteringProfile.

### CODE REVIEW

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp [hunk 1]
```diff
@@ -199,6 +199,59 @@ namespace OpenMS
     }
   }
 
+
+  void IsotopeDistribution::estimateFromRNAWeight(double average_weight){
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: minor style issue: { on new line

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp [hunk 2]
```diff
@@ -199,6 +199,59 @@ namespace OpenMS
...
+      estimateFromWeightAndComp(average_weight,9.75,12.25,3.75,7,0,1);
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: minor style issue: we use space after comma

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp [hunk 3]
```diff
@@ -199,6 +199,59 @@ namespace OpenMS
...
+      names.push_back("C");
+      names.push_back("H");
+      names.push_back("N");
+      names.push_back("O");
+      names.push_back("S");
+      names.push_back("P");
+
+      //Averagine element count divided by averagine weight
+      vector<double> factors;
+      double monoTotal= (C*db->getElement("C")->getMonoWeight()+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: also space between operators

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 4]
```diff
@@ -452,13 +452,18 @@ namespace OpenMS
...
+
   double MultiplexFiltering::getAveragineSimilarity(const vector<double>& pattern, double m) const
+
   {
     // construct averagine distribution
     IsotopeDistribution distribution;
     vector<double> averagine_pattern;
     distribution.setMaxIsotope(pattern.size());
-    distribution.estimateFromPeptideWeight(m);
+    if (type_m_=="peptide") distribution.estimateFromPeptideWeight(m);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: style: space around comparison operators
use blocks of {} for multiple if else expressions

File: src/tests/class_tests/openms/source/IsotopeDistribution_test.cpp [hunk 5]
```diff
@@ -267,6 +267,15 @@ START_SECTION(void estimateFromPeptideWeight(double average_weight))
 	TEST_REAL_SIMILAR(iso.begin()->second, 0.00291426)
 END_SECTION
 
+START_SECTION(void estimateFromWeightAndComp(double average_weight, double C, double H, double N, double O, double S, double P))
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/IsotopeDistribution_test.cpp]: tests for DNA and RNA missing?

File: src/topp/FeatureFinderMultiplex.cpp [hunk 6]
```diff
@@ -226,6 +227,8 @@ class TOPPFeatureFinderMultiplex :
       defaults.setMinInt("missed_cleavages", 0);
       defaults.setValue("knock_out", "false", "Is it likely that knock-outs are present? (Supported for doublex, triplex and quadruplex experiments only.)", ListUtils::create<String>("advanced"));
       defaults.setValidStrings("knock_out", ListUtils::create<String>("true,false"));
+      defaults.setValue("type_m","peptide","The type of averagine to use, currently RNA, DNA or peptide");
```
Comment (timosachsenberg) [inline src/topp/FeatureFinderMultiplex.cpp]: maybe make this an advanced option (see knock_out parameter).
I would also use a more expressive name than "type_m" maybe "averagine_type"

**Comment by timosachsenberg** [comment]
there also are some files that need to be adapted for the python bindings:
see src/pyOpenMS/pxds/IsotopeDistribution.pxd

**Comment by timosachsenberg** [comment]
for reference: https://github.com/OpenMS/OpenMS/wiki/Coding-conventions

**Comment by poshul** [comment]
Yes, the composition of RNA and DNA is quite similar, the second and third tests, with larger average masses do have different values.

**Comment by timosachsenberg** [comment]
+1 looks good to me. let's wait for some more comments and then merge

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp [hunk 7]
```diff
@@ -199,6 +199,61 @@ namespace OpenMS
...
+      estimateFromWeightAndComp(average_weight, 9.75, 12.25, 3.75, 7, 0, 1);
+  }
+
+
+  void IsotopeDistribution::estimateFromDNAWeight(double average_weight)
+  {
+      estimateFromWeightAndComp(average_weight, 9.75, 12.25, 3.75, 6, 0, 1);
+  }
+
+  void IsotopeDistribution::estimateFromWeightAndComp(double average_weight, double C, double H, double N, double O, double S, double P)
```
Comment (lars20070) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: Why is `average_weight` needed? Is the weight not defined by the chemical composition?

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp [hunk 8]
```diff
@@ -199,6 +199,61 @@ namespace OpenMS
     }
   }
 
+
+  void IsotopeDistribution::estimateFromRNAWeight(double average_weight)
+  {
+      estimateFromWeightAndComp(average_weight, 9.75, 12.25, 3.75, 7, 0, 1);
```
Comment (lars20070) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: Could you please add a comment with a reference where these numbers are from?

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp (hunk 7 above)
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: the isotope distribution depends on the weight (same for peptides)

Comment (lars20070) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: This method is not used in `MultiplexFiltering::getAveragineSimilarity`? Where is it needed?

Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: see 4 lines above

**Comment by lars20070** [comment]
Could you please also add a TOPP test for both RNA and DNA case?

**Comment by lars20070** [comment]
@poshul  Thanks for the useful extension!

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 9]
```diff
@@ -452,13 +452,31 @@ namespace OpenMS
...
+    {
+        distribution.estimateFromRNAWeight(m);
+    }
+    else if (averagine_type_ == "DNA")
+    {
+        distribution.estimateFromDNAWeight(m);
+    }
+    else
+    {
+        distribution.estimateFromPeptideWeight(m); //TODO: Deal with this intelligently
```
Comment (lars20070) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: What do you have in mind? - Can you not simply throw a `InvalidParameter` exception?
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/CONCEPT/Exception.h#L348

File: src/openms/source/CHEMISTRY/IsotopeDistribution.cpp (hunk 7 above)
Comment (lars20070) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: Thanks.

Comment (lars20070) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: The `S` parameter is only needed for potentially sulfur-modified nucleotides?

Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/IsotopeDistribution.cpp]: It is even more general: see e.g. https://wikis.engrade.com/mahabiochem/elements_bio

**Comment by lars20070** [comment]
@poshul Perhaps you can send me one DNA and one RNA example file? I can then optimise the parameters and add two further tests for the TOPP tool [1] in a separate pull request. At some point, we can also add your complete workflow to our collection [2]. 

[1] https://github.com/OpenMS/OpenMS/blob/develop/src/tests/topp/CMakeLists.txt#L261
[2] https://github.com/OpenMS/OpenMS/wiki/Workflow-Collection

**Comment by poshul** [comment]
@lars20070: I will try to find a DNA file that I can share. I have RNA mzML's but they are all large, where should I sent them?

**Comment by timosachsenberg** [comment]
I think it is fine if you cut out a small region that contains a few features. e.g. 5 minutes and 100 Th.
You can use FileFilter to do so and send it to me lars (I send you his email in private).

**Comment by lars20070** [comment]
@poshul Can you please add two TOPP tool tests [1]? One with the small DNA example, one with the RNA example as @timosachsenberg suggested. Add the mzML, ini and featureXML here [2], and extend [3] with “TOPP_FeatureFinderMultiplex_3” and “TOPP_FeatureFinderMultiplex_4”.

[1] https://github.com/OpenMS/OpenMS/wiki/Developer-FAQ#add_topptest
[2] https://github.com/OpenMS/OpenMS/tree/develop/src/tests/topp
[3] https://github.com/OpenMS/OpenMS/blob/develop/src/tests/topp/CMakeLists.txt#L261

**Comment by lars20070** [comment]
+1

(after TOPP tests are added)

**Comment by poshul** [comment]
TOPP tests added.

**Comment by timosachsenberg** [comment]
+1

**Pull request merged**